src/PlusDataCollection/OpenHaptics/vtkPlusOpenHapticsDevice.cxx

```diff
@@ -0,0 +1,297 @@
+/*=Plus=header=begin======================================================
+Program: Plus
+Copyright (c) Laboratory for Percutaneous Surgery. All rights reserved.
+See License.txt for details.
+=========================================================Plus=header=end*/
+
+
+// Local includes
+#include "PlusConfigure.h"
+#include "vtkPlusDataSource.h"
+#include "vtkPlusOpenHapticsDevice.h"
+
+//HD includes
+#include <HDU/hduVector.h>
+#include <HDU/hduError.h>
+#include <HD/hd.h>
+
+// VTK includes
+#include <vtkMatrix4x4.h>
+#include <vtkTransform.h>
+
+// STL includes
+#include <fstream>
+#include <iostream>
+#include <sstream>
+
+vtkStandardNewMacro(vtkPlusOpenHapticsDevice);
+
+vtkPlusOpenHapticsDevice::vtkPlusOpenHapticsDevice()
+{
+  this->FrameNumber = 0;
+  this->isHapticDeviceInitialized = false;
+  this->DeviceHandle = -1;
+  this->RequirePortNameInDeviceSetConfiguration = true;
+  this->StartThreadForInternalUpdates = true;
+  this->AcquisitionRate = 20;
+  this->DeviceName = "PLUS";
+}
+
+vtkPlusOpenHapticsDevice::~vtkPlusOpenHapticsDevice()
+{
+  if (this->isHapticDeviceInitialized)
+  {
+    this->isHapticDeviceInitialized = false;
+  }
+}
+
+std::string vtkPlusOpenHapticsDevice::GetSdkVersion()
+{
+  std::stringstream ss;
+  ss << HD_VERSION_MAJOR_NUMBER << "." << HD_VERSION_MINOR_NUMBER << "." << HD_VERSION_BUILD_NUMBER;
+  return ss.str();
+}
+
+PlusStatus vtkPlusOpenHapticsDevice::Probe()
+{
+  if (this->isHapticDeviceInitialized)
+  {
+    LOG_ERROR("vtkPlusOpenHapticsDevice::Probe should not be called while the device is already initialized");
+    return PLUS_FAIL;
+  }
+  if (this->InternalConnect())
+  {
+    this->InternalDisconnect();
+    LOG_INFO("vtkPlusOpenHapticsDevice::Probe found OpenHaptics Device");
+    return PLUS_SUCCESS;
+  }
+  else
+  {
+    LOG_ERROR("vtkPlusOpenHapticsDevice::Probe failed to connect to device");
+    return PLUS_FAIL;
+  }
+
+}
+
+PlusStatus vtkPlusOpenHapticsDevice::InternalUpdate()
+{
+  //schedule the internal callback - must be done each time device is polled
+  //the scheduler is intended to manage multiple clients requesting data from the same device
+  hdScheduleSynchronous(positionCallback, this, HD_MAX_SCHEDULER_PRIORITY);
+
+  return PLUS_SUCCESS;
+}
+
+PlusStatus vtkPlusOpenHapticsDevice::InternalConnect()
+{
+  if (this->isHapticDeviceInitialized)
+  {
+    LOG_DEBUG("Already connected - " << this->DeviceName);
+    return PLUS_SUCCESS;
+  }
+
+  if (this->InputChannels.empty())

```

Comment:
This type of check is better placed in NotifyConfigured, so the system will fail sooner, rather than try to connect to some devices before noticing this failure. Some devices have long connect times, and NotifyConfigured provides an early exit for configuration problems.